tests/middleware/test_base.py

```diff
@@ -84,3 +83,41 @@ def homepage(request):
     response = client.get("/homepage")
     assert response.text == "Homepage"
     assert response.headers["Custom"] == "Example"
+
+def test_state_data_across_multiple_middlewares():
```

Comment:
I'd suggest that we could simplify this down to a simpler style of test that:

* Has a single middleware, which sets some state on the request.
* Has a view, which returns `JSONResponse(dict(request.state))` 

That'd be sufficient for ensuring that `.state` is persisted across two different request instances.

Reply:
When I moved `"state"` init inside `@property def state():`, my test bombed in a meaningful way:

The middle state var called `bar` was not persisted.

Stacktrace:

```
_____________________________________________________________________________ test_state_data_across_multiple_middlewares ______________________________________________________________________________

    def test_state_data_across_multiple_middlewares():
        expected_value1 = "foo"
        expected_value2 = "bar"
    
        class aMiddleware(BaseHTTPMiddleware):
            async def dispatch(self, request, call_next):
                request.state.foo = expected_value1
                response = await call_next(request)
                return response
    
        class bMiddleware(BaseHTTPMiddleware):
            async def dispatch(self, request, call_next):
                request.state.bar = expected_value2
                response = await call_next(request)
                response.headers["X-State-Foo"] = request.state.foo
                return response
    
        class cMiddleware(BaseHTTPMiddleware):
            async def dispatch(self, request, call_next):
                response = await call_next(request)
                response.headers["X-State-Bar"] = request.state.bar
                return response
    
        app = Starlette()
        app.add_middleware(aMiddleware)
        app.add_middleware(bMiddleware)
        app.add_middleware(cMiddleware)
    
        @app.route("/")
        def homepage(request):
            return PlainTextResponse("OK")
    
        client = TestClient(app)
>       response = client.get("/")

tests/middleware/test_base.py:122: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
venv/lib/python3.7/site-packages/requests/sessions.py:546: in get
    return self.request('GET', url, **kwargs)
starlette/testclient.py:421: in request
    json=json,
venv/lib/python3.7/site-packages/requests/sessions.py:533: in request
    resp = self.send(prep, **send_kwargs)
venv/lib/python3.7/site-packages/requests/sessions.py:646: in send
    r = adapter.send(request, **kwargs)
starlette/testclient.py:238: in send
    raise exc from None
starlette/testclient.py:235: in send
    loop.run_until_complete(self.app(scope, receive, send))
/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py:584: in run_until_complete
    return future.result()
starlette/applications.py:133: in __call__
    await self.error_middleware(scope, receive, send)
starlette/middleware/errors.py:122: in __call__
    raise exc from None
starlette/middleware/errors.py:100: in __call__
    await self.app(scope, receive, _send)
starlette/middleware/base.py:25: in __call__
    response = await self.dispatch_func(request, self.call_next)
tests/middleware/test_base.py:109: in dispatch
    response.headers["X-State-Bar"] = request.state.bar
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <starlette.requests.State object at 0x10ab52e48>, key = 'bar'

    def __getattr__(self, key: typing.Any) -> typing.Any:
        if key in self._state:
            return self._state[key]
        else:
            raise AttributeError(
>               "'{}' object has no attribute '{}'".format(self.__class__.__name__, key)
            )
E           AttributeError: 'State' object has no attribute 'bar'

starlette/requests.py:33: AttributeError
```